pydantic_ai_slim/pydantic_ai/_pydantic.py

```diff
@@ -150,13 +152,16 @@ def function_schema(  # noqa: C901
     )
     # PluggableSchemaValidator is api compatible with SchemaValidator
     schema_validator = cast(SchemaValidator, schema_validator)
-    json_schema = GenerateJsonSchema().generate(schema)
+    json_schema = schema_generator().generate(schema)
 
     # workaround for https://github.com/pydantic/pydantic/issues/10785
-    # if we build a custom TypeDict schema (matches when `single_arg_name is None`), we manually set
+    # if we build a custom TypedDict schema (matches when `single_arg_name is None`), we manually set
     # `additionalProperties` in the JSON Schema
     if single_arg_name is None:
-        json_schema['additionalProperties'] = bool(var_kwargs_schema)
+        if schema_generator is GenerateJsonSchema:
+            # If you are overriding the GenerateJsonSchema, you should handle this yourself
+            # TODO: It might make sense to instead use a custom subclass of GenerateJsonSchema that does this..
```

Comment:
I guess this was the "workaround" described above. I feel like it would be better if we didn't rely on this hard-coding here, it makes it impossible to remove the `additionalProperties` field if you want it. I would suggest we have a `pydantic-ai`-specific subclass of `GenerateJsonSchema` as the default, and that way we can make this inclusion of `additionalProperties` handled there and not here. So you can then change that behavior as a user if you want to.

Reply:
Seems better to have the subclass, yep.